python/ql/lib/semmle/python/frameworks/data/internal/ApiGraphModelsSpecific.qll

```diff
@@ -41,8 +45,59 @@ predicate hasImplicitTypeModel(string type, string otherType) { none() }
 bindingset[type, path]
 API::Node getExtraNodeFromPath(string type, AccessPath path, int n) { none() }
 
+/**
+ * Holds if `type` = `typePath`+`suffix` and `suffix` is either empty or "!".
+ */
+bindingset[type]
+private predicate parseType(string type, string typePath, string suffix) {
+  exists(string regexp |
+    regexp = "([^!]+)(!|)" and
```

Comment:
Is `(!|)` different from `(!?)` somehow? (Apart from the latter expressing some level of surprise. 😅)

Reply:
I must admit that I just copied the Ruby one...I kind of like the surprised version :-)